File: Pharmacy-Persistence/src/de/java/domain/OrderState.java
package de.java.domain;

public enum OrderState {
  OPEN {
    public OrderState next() { return POSTING; }
  },
  POSTING {
    public OrderState next() { return ORDERED; }
    public boolean mayBeCancelled() { return true; }
    public OrderState cancel() { return CANCELLED; }
  },
  ORDERED {
    public OrderState next() { return FINISHED; }
  },
  FINISHED {
    public OrderState next() { throw new IllegalOrderStatusTransitionException(); }
    public boolean isProceedable() { return false; }
  },
  CANCELLED {
    public OrderState next() { throw new IllegalOrderStatusTransitionException(); }
    public boolean isProceedable() { return false; }
  };

  public abstract OrderState next();

  public boolean isProceedable() {
    return true;
  }

  public boolean mayBeCancelled() {
    return false;
  }

  public OrderState cancel() {
    throw new IllegalOrderStatusTransitionException();
  }

}

File: DotNET/BusinessLayer/Data/OrderState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Pharmacy.BusinessLayer.Data
{
    public enum OrderState
    {
        Open, Posting, Ordered, Finished, Cancelled
    }

    public static class OrderStateExtension
    {
        public static OrderState Next(this OrderState currentState)
        {
            switch (currentState)
            {
                case OrderState.Open:
                    return OrderState.Posting;
                case OrderState.Posting:
                    return OrderState.Ordered;
                case OrderState.Ordered:
                    return OrderState.Finished;
                default:
                    // to have a reasonable default behaviour (instead of throwing an exception)
                    return currentState;
            }
        }

        public static Boolean Proceedable(this OrderState state)
        {
            return !(state == OrderState.Finished || state == OrderState.Cancelled);
        }

        public static OrderState Cancel(this OrderState state)
        {
            if (state != OrderState.Posting)
                throw new ArgumentException("Order can only be cancelled in state " + OrderState.Posting);
            return OrderState.Cancelled;
        }

        public static Boolean Cancellable(this OrderState state)
        {
            return state == OrderState.Posting;
        }
    }
}
